### ISSUE TITLE
Add type annotations

### ISSUE THREAD

**Issue by matangover** [librosa/librosa#1564]
I would like to have code completions and type checking for code that uses librosa. Adding type annotations will significantly improve the coding experience for users who use type checkers / typing-based completion engines.

Currently some type checkers like Pylance are able to infer some types, however this is often wrong and leads to false positives (errors on working code) or false negatives (missing errors on invalid code). Other type checkers like Mypy cannot do anything with code that uses librosa, and must treat it as 'Any'.

**Comment by bmcfee** [comment]
Sure, this would be great to have. (Especially for the 1.0 release.)

It's also an enormous undertaking that I don't have time for in the short term.  Help on this would be much appreciated.

**Comment by matangover** [comment]
Cool. I know some tools that make it easier to add type annotations for existing code using static inference / runtime introspection. If I have time I'll try running one of them and inspecting the results. How exhaustive are the existing tests in `tests/`? Could they be sufficient to verify that the generated typings are correct (by checking that the tests pass a type check)? Otherwise, do you have other sources of "a bunch of code that uses librosa" which can be used for that purpose?

**Comment by bmcfee** [comment]
> How exhaustive are the existing tests in `tests/`? Could they be sufficient to verify that the generated typings are correct (by checking that the tests pass a type check)?

We aim for full code path coverage, though of course that's not the same thing as being exhaustive with respect to types since the input types are generally not bound.  I do think that the test suite is probably your best bet though, as most of the external code that's out there will only use a small subset of the library.

I think the bigger headache here will be going back through and implementing proper [numpy typing](https://numpy.org/devdocs/reference/typing.html#number-precision) where appropriate (which will bump our min numpy up to 1.20).

**Comment by matangover** [comment]
Gotcha. Precise numpy typing is definitely nice to have but we could also do that as a second step

**Comment by bmcfee** [comment]
Note: #1577 bumped our minimum python up to 1.20, so consider the numpy typing issue unblocked.

**Comment by bmcfee** [comment]
Following up on this, #1579 adds stub files (pyi) for the top-level package and submodules (anything with a `__init__.py` gets a `__init__.pyi`).

While hacking this out, I also went down the rabbit hole of generating full stub declarations for all modules using some automated tooling ([stubgen](https://mypy.readthedocs.io/en/stable/stubgen.html)).  This didn't by any means get the job done, but it at least roughed out the function declarations enough that we could go in and insert proper types.  This seems like the easiest route to follow in implementing type hints for this PR.

Note: it will still be a big undertaking.

**Comment by bmcfee** [comment]
Nice - thanks for looking into it.  I think we might still want / need stub files in some situations though, either because of dependency types from upstream packages, or compiled extensions.  Not yet sure how numba JIT functions will enter into this, but it wouldn't surprise me if stubs are necessary there too.

**Comment by matangover** [comment]
I added annotations to librosa, please check the [typing branch](https://github.com/matangover/librosa/tree/typing) of my fork. I should open a PR at some point after some more clean up

How I did it:
1. Wrote and ran a "docstrings to annotations" script
2. Type checked the test suite and adjusted typings to resolve type errors reported by mypy and Pylance. Also had to make small adjustments to the tests like adding casts where typings are ambiguous etc, but I tried to keep this to a minimum and make typings more specific by adding overloads etc.
3. Ran `pyright --verifytypes` to make sure public exported symbols are fully typed (or at least that presence of `Any` is explicit). Added typings where necessary

Note that the librosa code itself does currently generate many type errors, but this is out of scope for this change. My goal is just to add annotations to the public interface to make it nicer for users.

Caveats:
1. I only tested with mypy and Pylance (Pyright). I think they are the most popular checkers
2. `ndarray` was made generic only in numpy>=1.21 ([PR](https://github.com/numpy/numpy/pull/17719)). So I can't specify array dtype without upgrading numpy. This can be addressed in the future
4. Runtime dependency of typing_extensions is added to support `Literal` in Python<3.8
5. Some functions were hard to type accurately, e.g.:
    * Return types varying based on arguments, e.g. `dtw` needs 9 overloads to type accurately I believe (3 ** 2). I didn't take the time to really do this properly for all functions, it can be improved in the future.
    * Functions that take either scalar or sequence. e.g. with `note_to_midi` it was hard because `str` is a subclass of `Iterable[str]`, so I had to compromise on `List[str]`...
    * Functions that involve math operations on either Python scalars, numpy scalars, or numpy arrays (e.g. `hz_to_midi`) and use `asanyarray` on the input -- which might result in a 0d array which is then cast by numpy to a numpy scalar ...
9. librosa uses `from .some_module import *` to re-export symbols. This is supported by Pyright and mypy. However in mypy with `--strict` enabled (which enables `--no-implicit-reexport`), this is disabled. We can solve this by adding `__all__` or using stub file


The test suite is now free of librosa typing errors. There are still some errors reported such as:
1. Accessing private members from the tests (e.g. `__reassign_frequencies`), might want to rename double underscore to single underscore
2. Missing annotations of other libraries (matplotlib...).
3. Errors resolved by a [change](https://github.com/numpy/numpy/pull/22359) already in numpy main
4. [Bug](https://github.com/microsoft/pyright/issues/4000) in pyright

**Comment by bmcfee** [comment]
Wow, thanks for this!  There's certainly a lot to review here, but so far it looks great.

Some responses to your caveats:



> I only tested with mypy and Pylance (Pyright). I think they are the most popular checkers

:+1:



> 2\. `ndarray` was made generic only in numpy>=1.21 ([PR](https://github.com/numpy/numpy/pull/17719)). So I can't specify array dtype without upgrading numpy. This can be addressed in the future

It'd be good to sort out a plan for this early on.  We do have runtime checks for dtypes in critical places (though not everywhere), and in general I prefer to be more permissive than not when it comes to dtypes.  So maybe this isn't a huge problem?



> Runtime dependency of typing_extensions is added to support `Literal` in Python<3.8

:+1:




> * Return types varying based on arguments, e.g. `dtw` needs 9 overloads to type accurately I believe (3 ** 2). I didn't take the time to really do this properly for all functions, it can be improved in the future.

Oof.  I imagine those points are few, but painful to deal with.



> * Functions that take either scalar or sequence. e.g. with `note_to_midi` it was hard because `str` is a subclass of `Iterable[str]`, so I had to compromise on `List[str]`...

O_o that's super weird, but makes sense in retrospect.  We can't be the first to run into this problem, I wonder if there's a more standard solution?



> * Functions that involve math operations on either Python scalars, numpy scalars, or numpy arrays (e.g. `hz_to_midi`) and use `asanyarray` on the input -- which might result in a 0d array which is then cast by numpy to a numpy scalar ...

Not sure I understand the problem here - can this not be handled by a sufficiently permissive union?



> 5\. librosa uses `from .some_module import *` to re-export symbols. This is supported by Pyright and mypy. However in mypy with `--strict` enabled (which enables `--no-implicit-reexport`), this is disabled. We can solve this by adding `__all__` or using stub file

This is actually not true anymore since we merged #1579.  You'll probably need to rebase this branch to make sure everything's still square.



> 6\. There's a balance between making the annotations correct and making them useful. E.g. `midi_to_hz` when given an `int` returns an `np.float64`, which is typing-wise incompatible with `float` (although in practice it is accepted almost anywhere float is accepted). If the user tries to pass it to a function which accepts `float` they would get a type error. On the other hand, if I annotate `midi_to_hz` as returning `float` it would be technically wrong and would mean that e.g. `midi_to_hz(X).dtype` would give a type error. This is a general problem with typing of numpy scalars... I chose to go with `np.floating`. The user can write `float(midi_to_hz(X))` to silence the error

Do you have a sense of how often this kind of thing occurs?  I'm not opposed to modifying the code if it generally preserves functionality and prevents contorting the type annotations too much.



> The test suite is now free of librosa typing errors. There are still some errors reported such as:

:exploding_head: nice!



> 1. Accessing private members from the tests (e.g. `__reassign_frequencies`), might want to rename double underscore to single underscore

Yes, I'd be fine with that in general.  Some of the functions probably should stay "really private" though, especially when it comes to numba helpers and the like.  I think we'll need to take these one at a time.



> Missing annotations of other libraries (matplotlib...).

This raises an interesting point. MPL has generally delayed this feature until the upstream packages stabilize, and that makes a lot of sense.  I guess the thing to do here is try to get a clear picture of what upstream API's we directly expose / interoperate with and see how much we'd be held back by waiting for them to implement type annotations.  My expectation is "not very much", but I could be missing something.

**Comment by bluenote10** [comment]
> Functions that involve math operations on either Python scalars, numpy scalars, or numpy arrays (e.g. hz_to_midi) and use asanyarray on the input -- which might result in a 0d array which is then cast by numpy to a numpy scalar ...

Perhaps we just need a few common type aliases defined in a central place like (ignoring dtype annotation for now)? 

```
ArrayLike = Union[np.ndarray, Sequence[float], Sequence[complex]]
ScalarOrArrayLike = Union[float, complex, ArrayLike]
```

> Functions that take either scalar or sequence. e.g. with note_to_midi it was hard because str is a subclass of Iterable[str], so I had to compromise on List[str]...

Shouldn't `note: Union[str, Iterable[str]]` do the job nonetheless? All that matters is that both `librosa.note_to_midi('C')` and `librosa.note_to_midi(['C', 'E', 'G'])` type checks, right? If `Iterable` really must be avoided we could at least use `Sequence[str]` to be a bit broader and express immutability.

**Comment by matangover** [comment]
> Wow, thanks for this!

You're welcome :)

> I wonder if there's a more standard solution

Nope, as far as I can see [other libraries](https://github.com/pandas-dev/pandas/issues/47232) encountered this too and List seems like the way to go. This is a general problem, see [issue](https://github.com/python/typing/issues/256). If you try to write two overloads, one accepting `str` and other accepting `Iterable[str]`, you get type error "overload signatures overlap with incompatible return types", which makes sense because otherwise you could do:
```python
str_as_iterable: Iterable[str] = "my_string"
typed_as_array_but_is_actually_float = librosa.note_to_midi(str_as_iterable)
```

> Not sure I understand the problem here - can this not be handled by a sufficiently permissive union?

It's not a problem, just was a challenge

> This is actually not true anymore since we merged https://github.com/librosa/librosa/pull/1579

Oh, great

> Do you have a sense of how often this kind of thing occurs?

Umm, can't say...


> see how much we'd be held back by waiting for them to implement type annotations

I don't think there's any good reason to wait with publishing typings. The small number of functions that return/accept matplotlib types will simply show up as Any. Also, matplotlib, sklearn, scipy stubs are available by default to users of Pylance, and other type checkers can download [Microsoft's stubs](https://github.com/microsoft/python-type-stubs).

> Perhaps we just need a few common type aliases defined in a central place like (ignoring dtype annotation for now)?

Specifically ArrayLike is already defined by numpy. And I added librosa._typing for other common definitions.

> All that matters is that both librosa.note_to_midi('C') and librosa.note_to_midi(['C', 'E', 'G']) type checks, right?

Well, no... first of all `Union[str, Iterable[str]]` is equivalent to `Iterable[str]`... But the real point is that you want to avoid Unions in return type if possible because they cause false positives when the return value is used. E.g. you want the type system to know that `note_to_midi('C')` returns a float and `note_to_midi(['C', 'E', 'G'])` returns an array. That's why overloads make the typing much better from the user's point of view.

> we could at least use Sequence[str] 

No, str inherits from Sequence[str] in `builtins.pyi`

**Comment by matangover** [comment]
I rebased on the lazy load branch, and now `mypy --no-implicit-reexport tests` no longer throws type errors, good job adding the reexport in the pyi files!

**Comment by bluenote10** [comment]
> Well, no... first of all Union[str, Iterable[str]] is equivalent to Iterable[str]... But the real point is that you want to avoid Unions in return type if possible because they cause false positives when the return value is used.

Sorry, having only skimmed the code, I thought you're only talking about the input side. Fully agreed that overloading is preferred if return types are involved.

> No, str inherits from Sequence[str] in builtins.pyi

Ah yeah, that's a pity. What about overloading for `IteratorLike[str]` defined as ([try on playground](https://mypy-play.net/?mypy=latest&python=3.10&gist=3a6fafd3bb77305093b23994f6ffab4e)):

```python
from typing import TypeVar, Union, List, Tuple, Generator, overload

T = TypeVar("T")

IteratorLike = Union[List[T], Tuple[T, ...], Generator[T, None, None]]

@overload
def f(s: IteratorLike[str]) -> List[float]:
    ...
    
@overload
def f(s: str) -> float:
    ...

def f(s: Union[str, IteratorLike[str]]) -> Union[float, List[float]]:
    ...

# Seems to do the job
reveal_type(f("C"))
reveal_type(f(["C", "D", "E"]))
reveal_type(f(("C", "D", "E")))
reveal_type(f("C" for _ in range(3)))
```

which should basically cover all the majority of practically relevant usages?

**Comment by matangover** [comment]
Yes that could definitely work

**Comment by bmcfee** [comment]
:wave: just to bring this conversation back around, I'm wondering what it would take to get mypy to pass cleanly on the main source tree (not tests) at this point?  I've been trying to piece this out locally over the past few days, and I'm getting a bit lost between things that are either bad/broken stubs from upstream and things that might just be incorrect on our side.

**Comment by matangover** [comment]
Running mypy in non-strict mode gets currently ~200 errors. You'd have to go one by one and resolve them. Here's the `mypy.ini` I'm using:

```ini
[mypy]
check_untyped_defs = True
allow_redefinition = True
files = librosa/

; Ignore libraries for which no type annotations are available.
[mypy-scipy.*]
ignore_missing_imports = True

[mypy-soundfile.*]
ignore_missing_imports = True

[mypy-audioread.*]
ignore_missing_imports = True

[mypy-sklearn.*]
ignore_missing_imports = True

[mypy-matplotlib.*]
ignore_missing_imports = True

[mypy-numba.*]
ignore_missing_imports = True

[mypy-decorator.*]
ignore_missing_imports = True

[mypy-joblib.*]
ignore_missing_imports = True

[mypy-pooch.*]
ignore_missing_imports = True

[mypy-lazy_loader.*]
ignore_missing_imports = True

[mypy-soxr.*]
ignore_missing_imports = True

[mypy-msgpack.*]
ignore_missing_imports = True

```

**Comment by bmcfee** [comment]
(Reopening this issue until mypy passes cleanly)

**Comment by bmcfee** [comment]
> Running mypy in non-strict mode gets currently ~200 errors. You'd have to go one by one and resolve them.

Yes, this is the part I was asking about.  From a quick glance, it seems like most of them are either errors in propagating out of numpy.typing, or incorrect inferences due to limitations of static analysis (eg branching paths depending on types at runtime).  It's not obvious to me how much work there actually is to do here, or if it's just going to amount to adding a ton of `# type: ignore` comments because mypy isn't as smart as we'd like.  Any insight here would be appreciated.

**Comment by matangover** [comment]
Yes that's right. It's a big effort to try and fix all errors at once. Type checking is useful - e.g. for finding subtle bugs (like the FJSFormatter case) - but applying it all at once to a codebase that was written without typing in mind can be a real time sucker. That's why mypy encourages the 'gradual typing' use case. My personal suggestion is to make sure new code is typed correctly from now on (e.g. if you're using VS Code then just enabling Pylance type checking or using mypy-vscode). Regarding existing code, probably should be done gradually.

**Comment by matangover** [comment]
Most of the work adding typing would be 'helping' the type checker know things that you already know (or assume) are correct. E.g. if you look at `TimeFormatter.__call__` you'd see Pyright giving an error:

<img width="716" alt="image" src="https://user-images.githubusercontent.com/5028956/208112114-980eb2ea-b9c4-4678-97b2-74fdfde06ab8.png">

And you'd then have to uncomment the `else: assert False` that I added, to tell the type checker that you "know" you have covered all cases in those if branches.

**Comment by bmcfee** [comment]
> That's why mypy encourages the 'gradual typing' use case. My personal suggestion is to make sure new code is typed correctly from now on (e.g. if you're using VS Code then just enabling Pylance type checking or using mypy-vscode). Regarding existing code, probably should be done gradually.

Sure, I appreciate that point.  However, if we don't have this implemented in a CI pipeline, it's going to be a lot more work to maintain typing on new contributions.  I'd rather just get this all sorted out up front and then not have to worry about it.  I'm not going to rely on contributors to use a specific text editor, and given the variance we've already noticed between different checkers (pyright, mypy), I wouldn't trust that anyway.





> And you'd then have to uncomment the `else: assert False` that I added, to tell the type checker that you "know" you have covered all cases in those if branches.

What a strange pattern!  Is this kind of thing documented somewhere?  As a general rule, I avoid putting assertions in live code (exceptions are always more informative), but I understand the logic here.  It will definitely uglify a lot of code to add these extra cases, and we'll likely need to inject some coverage skipping in our test suite, but I guess it's worth it to get mypy to shut up? :laughing:

**Comment by matangover** [comment]
>  Is this kind of thing documented somewhere?

The following docs are relevant (partially overlapping):
- Pyright docs (e.g. [this](https://github.com/microsoft/pyright/blob/main/docs/type-concepts.md) and [this](https://github.com/microsoft/pyright/blob/main/docs/type-inference.md) page)
- Mypy docs (e.g. [this page](https://mypy.readthedocs.io/en/stable/type_inference_and_annotations.html))
- https://typing.readthedocs.io/en/latest/

> I guess it's worth it to get mypy to shut up?

Personally I don't think it's that ugly actually, it clarifies an assertion that was previously implicit (that the if/elif branches cover all possible cases).

> CI pipeline

OK, if that's what you want then one option is to set up mypy on the CI, and add a [file-global](https://mypy.readthedocs.io/en/stable/common_issues.html#ignoring-a-whole-file) `#type: ignore` comment for all files that have type errors, and then remove them gradually

**Comment by bmcfee** [comment]
> Personally I don't think it's that ugly actually, it clarifies an assertion that was previously implicit (that the if/elif branches cover all possible cases).

Sure, but a comment could also do that without adding code that's dead on arrival.





> OK, if that's what you want then one option is to set up mypy on the CI, and add a [file-global](https://mypy.readthedocs.io/en/stable/common_issues.html#ignoring-a-whole-file) `#type: ignore` comment for all files that have type errors, and then remove them gradually

This is actually what I've been doing in my local dev environment.  Once the dust settles I'll add the CI workflow and hopefully not have to ignore files wholesale.

**Comment by bmcfee** [comment]
Just to follow up with a concrete example.  mypy's doing the following:

```
librosa/beat.py:348: error: Slice index must be an integer or None  [misc]
```

The offending code block is here:

https://github.com/librosa/librosa/blob/8c4c1958888bb1c9a81af317106864af50dcf654/librosa/beat.py#L345-L348

Digging into it, it appears that `ndarray.argmax` is annotated to return `Array` (in 1.23.5 anyway), though if no axis is provided, it will return `np.int64`.  (Probably this is a clue that the numpy types are incorrect / could be overloaded to be more precise.)

The obvious fix here is to explicitly cast (not type hint!) the argmax result to int prior to use.  I'd be okay with this in rare circumstances, but I fear that this kind of thing is going to be the norm, not the exception, when dealing with numpy results.  The result is going to be a mess of type-cast code reminiscent of C++.  I'd really like to avoid that situation if possible.

**Comment by matangover** [comment]
If you `reveal_type(max_idx)` (or hover over it with Pylance) you'll see that it is correctly inferred to be `np.signedinteger`. (`argmax` has three overloads which you can see by using Pylance's Go to Declaration). So the problem is that mypy is overly restrictive about indexing, see:

https://github.com/python/mypy/issues/2410#issuecomment-863091868

**Comment by matangover** [comment]
I've been trying to help with adding some annotations around librosa code and just wanted to report happily that I found a small bug with type checking :) It's a missing parameter validation:
<img width="1005" alt="image" src="https://user-images.githubusercontent.com/5028956/208210618-bda7baba-7df3-4eff-ad49-7f79b8e1120c.png">

Solved by adding the missing else clause:
<img width="798" alt="image" src="https://user-images.githubusercontent.com/5028956/208210779-2e74e09f-9926-4748-a7a6-619db228a703.png">

**Comment by bmcfee** [comment]
@matangover I'm still hacking through mypy errors, and wondering about the ScalarOrSequence type.  It seems like this is what the numpy [ArrayLike](https://numpy.org/doc/1.20/reference/typing.html#numpy.typing.ArrayLike) type is for - is there a reason we didn't use this originally?

**Comment by matangover** [comment]
ArrayLike doesn't let you limit the type of scalar you accept
